src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1260 @@
+/**
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) Bao Project and Contributors. All rights reserved.
+ */
+
+#include <vaplic.h>
+#include <vm.h>
+#include <cpu.h>
+#include <emul.h>
+#include <mem.h>
+#include <interrupts.h>
+#include <arch/csrs.h>
+
+#define APLIC_MIN_PRIO      (0xFF)
+#define UPDATE_ALL_HARTS    (-1)
+#define MASK_INTP_ZERO      (0xFFFFFFFE)
+
+#define BIT32_SET_INTP(reg, intp_id) (reg[intp_id/32] =\
+                                      bit32_set(reg[intp_id/32], intp_id%32))
+#define BIT32_GET_INTP(reg, intp_id) ((bit32_get(reg[intp_id/32], intp_id%32)\
+                                      != 0) ? 1U : 0U)
+#define BIT32_CLR_INTP(reg, intp_id) (reg[intp_id/32] =\
+                                      bit32_clear(reg[intp_id/32], intp_id%32))
+
+#define ADDR_INSIDE_RANGE(addr, start, end)\
+                         (addr >= offsetof(struct aplic_global_hw, start) &&\
+                          addr  < offsetof(struct aplic_global_hw, end))
+
+#define GET_HART_INDEX(vcpu, intp_id) ((vaplic_get_target(vcpu, intp_id) >> \
+                                        APLIC_TARGET_HART_IDX_SHIFT) & \
+                                        APLIC_TARGET_HART_IDX_MASK)
+
+#define INTP_VALID(intp_id) (intp_id != 0 && intp_id < APLIC_MAX_INTERRUPTS)
```

Comment:
I forget to replace the conditions with this macro. I will keep it since I believe it helps to get a cleaner code. I noticed that you leave a comment to replace a macro function for an inline function. I do the same here. Commit 30829c33a6084d34476ecf779eb76d0d6b0a93bf